sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -625,6 +626,60 @@ def _update_loop_specs(
       dsl_component_spec.pop_input_from_component_spec(group_component_spec,
                                                        loop_argument_name)
 
+  def _populate_metrics_in_dag_outputs(
+      self,
+      ops: List[dsl.ContainerOp],
+      op_to_parent_groups: Dict[str, List[str]],
+      pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+  ) -> None:
+    """Populates metrics artifacts in dag outputs.
+
+    Args:
+      ops: The list of ops that may produce metrics outputs.
+      op_to_parent_groups: The dict of op name to parent groups. Key is the op's
+        name. Value is a list of ancestor groups including the op itself. The
+        list of a given op is sorted in a way that the farthest group is the
+        first and the op itself is the last.
+      pipeline_spec: The pipeline_spec to update in-place.
+    """
+    for op in ops:
+      op_task_spec = getattr(op, 'task_spec',
+                             pipeline_spec_pb2.PipelineTaskSpec())
+      op_component_spec = getattr(op, 'component_spec',
+                                  pipeline_spec_pb2.ComponentSpec())
+
+      # Get the component spec for all its parent groups.
+      parent_groups_component_specs = [pipeline_spec.root]
+      # skip the op itself and the root group which cannot be retrived via name.
+      for group_name in op_to_parent_groups[op.name][1:-1]:
+        component_name = dsl_utils.sanitize_component_name(group_name)
+        parent_groups_component_specs.append(
+            pipeline_spec.components[component_name])
+      # Reverse the order to make the farthest group in the end.
+      parent_groups_component_specs.reverse()
```

Comment:
If we traversed `op_to_parent_groups` in another way maybe we can save this call?

Reply:
Thanks Jiaxiao. I thought about that but it feels to me having an explicit traverse might be more readable. The list here likely won't be too large, so the extra cost of reserving should be acceptable. 
That said, I might change this in a later PR when I touch things around. Just keep it as-is for now so that I don't need another lgtm on this PR. :)